openmeter/subscription/connector.go

```diff
@@ -0,0 +1,384 @@
+	"github.com/openmeterio/openmeter/pkg/clock"
+	"github.com/openmeterio/openmeter/pkg/convert"
+	"github.com/openmeterio/openmeter/pkg/currencyx"
+	"github.com/openmeterio/openmeter/pkg/framework/transaction"
+	"github.com/openmeterio/openmeter/pkg/models"
+)
+
+type NewSubscriptionRequest struct {
+	Namespace  string
+	ActiveFrom time.Time
+	CustomerID string
+	Currency   currencyx.Code
+
+	Plan PlanRef
+
+	// The SubscriptionItem customizations compared to the plan
+	ItemCustomization []Patch
+
+	// TODO: Add discounts, either separately or as part of the patch language
+}
+
+type Connector interface {
+	Command
+	Query
+}
+type commandAndQuery struct {
+	repo Repository
+	// managers
+	entitlementManager  EntitlementManager
+	priceManager        PriceManager
+	subscriptionManager SubscriptionManager
+	// connectors
+	priceConnector  price.Connector
+	customerService customer.Service
+	// adapters
+	planAdapter        PlanAdapter
+	entitlementAdapter EntitlementAdapter
+	// framework
+	transactionManager transaction.Creator
+}
+
+func NewConnector(
+	repo Repository,
+	// managers
+	entitlementManager EntitlementManager,
+	priceManager PriceManager,
+	subscriptionManager SubscriptionManager,
+	// connectors
+	priceConnector price.Connector,
+	customerService customer.Service,
+	// adapters
+	planAdapter PlanAdapter,
+	entitlementAdapter EntitlementAdapter,
+	// framework
+	transactionManager transaction.Creator,
+) Connector {
+	return &commandAndQuery{
+		repo:                repo,
+		entitlementManager:  entitlementManager,
+		priceManager:        priceManager,
+		subscriptionManager: subscriptionManager,
+		priceConnector:      priceConnector,
+		customerService:     customerService,
+		planAdapter:         planAdapter,
+		entitlementAdapter:  entitlementAdapter,
+		transactionManager:  transactionManager,
+	}
+}
+
+type Command interface {
+	Create(ctx context.Context, req NewSubscriptionRequest) (Subscription, error)
+	Edit(ctx context.Context, subscriptionID models.NamespacedID, patches []Patch) (Subscription, error)
+	Cancel(ctx context.Context, subscriptionID string, at time.Time) (Subscription, error)
+}
+
+func (c *commandAndQuery) Create(ctx context.Context, req NewSubscriptionRequest) (Subscription, error) {
+	def := Subscription{}
+	currentTime := clock.Now()
+
+	// Fetch the customer
+	cust, err := c.customerService.GetCustomer(ctx, customerentity.GetCustomerInput{
+		Namespace: req.Namespace,
+		ID:        req.CustomerID,
+	})
+	if err != nil {
+		return def, err
+	}
+
+	if cust == nil {
+		return def, fmt.Errorf("customer is nil")
+	}
+
+	// If user has a plan right now return an error
+	_, err = c.repo.GetCustomerSubscription(ctx, models.NamespacedID{
+		ID:        req.CustomerID,
+		Namespace: req.Namespace,
+	})
+	if err != nil {
+		if _, ok := lo.ErrorsAs[*NotFoundError](err); !ok {
+			return def, err
+		}
+	} else {
+		return def, &models.GenericUserError{Message: "customer already has a subscription"}
```

Comment:
should this be a specific error?

Reply:
could be a Conflict instead of a BadRequest